fs_storage/models/fs_storage.py

```diff
@@ -0,0 +1,426 @@
+            warnings.simplefilter("always", DeprecationWarning)
+            warnings.warn(
+                fmt1.format(name=func1.__name__, reason=reason),
+                category=DeprecationWarning,
+                stacklevel=2,
+            )
+            warnings.simplefilter("default", DeprecationWarning)
+            return func1(*args, **kwargs)
+
+        return new_func1
+
+    return decorator
+
+
+class FSStorage(models.Model):
+    _name = "fs.storage"
+    _backend_name = "fs_storage"
+    _description = "FS Storage"
+
+    __slots__ = ("__fs", "__odoo_storage_path")
+
+    def __init__(self, env, ids=(), prefetch_ids=()):
+        super().__init__(env, ids=ids, prefetch_ids=prefetch_ids)
+        self.__fs = None
+        self.__odoo_storage_path = None
+
+    name = fields.Char(required=True)
+    code = fields.Char(
+        required=True,
+        help="Technical code used to identify the storage backend into the code."
+        "This code must be unique. This code is used for example to define the "
+        "storage backend to store the attachments vi the configuration parameter "
+        "'ir_attachment.storage.force.database' when the module 'fs_attachment' "
+        "is installed.",
+    )
+    protocol = fields.Selection(
+        selection="_get_protocols",
+        required=True,
+        default="odoofs",
+        help="The protocol used to access the content of filesystem.\n"
+        "This list is the one supported by the fsspec library (see "
+        "https://filesystem-spec.readthedocs.io/en/latest). A filesystem protocol"
+        "is added by default and refers to the odoo local filesystem.\n"
+        "Pay attention that according to the protocol, some options must be"
+        "provided through the options field.",
+    )
+    protocol_descr = fields.Text(
+        compute="_compute_protocol_descr",
+    )
+    options = fields.Text(
+        help="The options used to initialize the filesystem.\n"
+        "This is a JSON field that depends on the protocol used.\n"
+        "For example, for the sftp protocol, you can provide the following:\n"
+        "{\n"
+        "    'host': 'my.sftp.server',\n"
+        "    'ssh_kwrags': {\n"
+        "        'username': 'myuser',\n"
+        "        'password': 'mypassword',\n"
+        "        'port': 22,\n"
+        "    }\n"
+        "}\n"
+        "For more information, please refer to the fsspec documentation:\n"
+        "https://filesystem-spec.readthedocs.io/en/latest/api.html#built-in-implementations"
+    )
+
+    json_options = Serialized(
+        help="The options used to initialize the filesystem.\n",
+        compute="_compute_json_options",
+        inverse="_inverse_json_options",
+    )
+    directory_path = fields.Char(
+        help="Relative path to the directory to store the file"
+    )
+
+    # the next fields are used to display documentation to help the user
+    # to configure the backend
+    options_protocol = fields.Selection(
```

Comment:
I don't get why this field is usefull as we could use `protocol` to display the right `options_properties`.
Maybe I missed something, but it would avoid some code duplication between `_get_options_protocol` and `_get_protocols` and also make sure you display the description of the chosen protocol.
It is a bit strange to have to choose a protocol and then to have to choose again to see the right description in the available option tab.

Reply:
@florian-dacosta  With the lib you can nest the protocol. Therefore, the same field can be used to get access to the protocol documentation for the main protocol and when configuring nested protocols into the `options` field.